src/Features/Core/Completion/Providers/AbstractSymbolCompletionProvider.cs

```diff
@@ -297,18 +298,7 @@ private Task<AbstractSyntaxContext> GetOrCreateContext(Document document, int po
         {
             lock (s_cacheGate)
             {
-                Task<AbstractSyntaxContext> cachedContext;
-                if (s_cachedDocuments.TryGetValue(document, out cachedContext) && cachedContext != null)
-                {
-                    return cachedContext;
-                }
-            }
-
-            var context = CreateContext(document, position, cancellationToken);
-            lock (s_cacheGate)
-            {
-                s_cachedDocuments[document] = context;
-                return context;
+                return s_cachedDocuments.GetValue(document, d => CreateContext(d, position, cancellationToken));
```

Comment:
:bulb: `ConditionalWeakTable` is a thread-safe type. If multiple calls to `CreateContext` won't break things (i.e. the only downside would be duplicated/discarded work), consider removing the lock surrounding this call. This could be left as an item for future improvement.